corehq/apps/userreports/expressions/specs.py

```diff
@@ -816,7 +816,10 @@ def __call__(self, item, context=None):
     @ucr_context_cache(vary_on=('user_id',))
     def _get_groups(self, user_id, context):
         domain = context.root_doc['domain']
-        user = CommCareUser.get_by_user_id(user_id, domain)
+        try:
+            user = CommCareUser.get_by_user_id(user_id, domain)
+        except CouchUser.AccountTypeError:
+            user = None
```

Comment:
It's not clear to me that this was only ever meant to return mobile users. How are you sure of this?

Reply:
The main way is that it calls `CommCareUser` rather than the generic `CouchUser`, which would be used if either user could be expected, but additionally the two methods called on the returned user in the expressions inheriting from this class, `get_reporting_groups` and `get_case_sharing_groups` are both defined only for mobile users and do not exist in the `WebUser` class.

More generally, groups are really only a feature for mobile users (the UI will only let you add mobile workers to a group, and that is how they are documented https://confluence.dimagi.com/display/commcarepublic/Mobile+Worker+Groups), so these expressions would not make a lot of sense for web users, and other places where we need similar methods for web users we hardcode `[]` (https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/models.py#L182-L183)

Reply:
Ok, that seems reasonable. There are still two things hanging in my mind:

1. What exactly is the meaning of a `CouchUser.AccountTypeError` exception? As in, do we know the original author just "forgot" to catch that, or is it possible that they actually _meant_ to use `CouchUser.get_by_user_id(...)`?
2. The behavior implied by this code change doesn't correlate with the docstring on the `CouchUser.get_by_user_id()` [classmethod](https://github.com/dimagi/commcare-hq/blob/0868fb8e/corehq/apps/users/models.py#L1372-L1376):
    ```
    if domain is given, checks to make sure the user is a member of that domain
    returns None if there's no user found or if the domain check fails
    ```
    Looking at the code that follows, I can't tell if this docstring is just wrong, lacking information about raising an exception, or has a bug. Can `domain` here ever be `None`? If so, what was the original intention for that scenario?

To be clear: the reason for my scrutiny here is due to the nature of this change. It's deep in a subclass, and we already know that if it has a bug, it can easily go unnoticed (pillow retries are generally ignored until they grow in volume). I'm also concerned about avoiding regressions (even things that "worked", but perhaps weren't designed to work).

Reply:
> What exactly is the meaning of a CouchUser.AccountTypeError exception? 

`AccountTypeError` is raised when this method is called on a subclass of `CouchUser` (ie directly on `CommCareUser` or `WebUser`) and the returned user is of the wrong type. So in this case, the call is `CommCareUser.get_by_user_id` but the document with the id given is a `WebUser` instead.
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/users/models.py#L1381-L1382

> As in, do we know the original author just "forgot" to catch that, or is it possible that they actually meant to use CouchUser.get_by_user_id(...)?

As mentioned before, none of the methods called on the returned user exist for `WebUser` so returning the user would just cause an exception further down the line. Furthermore, none of the operations done on the return user make conceptual sense for a web user. They cannot be part of reporting or case sharing groups, so the only reasonable return value for those expressions is an empty list. We could in theory return the `WebUser` and either check the user type in each subclass or catch the `AttributeError` there, but this seemed both cleaner and safer.

> The behavior implied by this code change doesn't correlate with the docstring on the CouchUser.get_by_user_id() [classmethod](https://github.com/dimagi/commcare-hq/blob/0868fb8e/corehq/apps/users/models.py#L1372-L1376):

Can you explain a bit more what you mean here? I don't think the docstring is wrong, although agree it doesn't mention the exception. I can add another sentence mentioning how this exception works if you would like. Given that this check is 11 years old, and that it would not make sense to return a user of a different class (it would be confusing to call `CommCareUser.get_by_user_id` and have that method return a `WebUser` instead), I am fairly confident that raising a specific exception to the caller is both the intended and most reasonable behavior of this method.

> Can domain here ever (as in context.root_doc['domain']) be None? If so, what was the original intention for that scenario?

Assuming you mean in the UCR code calling here, domain can never be none, since in a UCR, the root doc (doc that we are inserting into the data source) will always be associated with a single domain. It should always be a form or a case, although we have looked into other UCR types in the past like locations, but all need to have a domain, or we would not be able to find the data source and configuration for them anyway.

There is also an assertion in this class to make sure that is the case https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/expressions/specs.py#L813

>  To be clear: the reason for my scrutiny here is due to the nature of this change. It's deep in a subclass, and we already know that if it has a bug, it can easily go unnoticed (pillow retries are generally ignored until they grow in volume). I'm also concerned about avoiding regressions (even things that "worked", but perhaps weren't designed to work).

I am not entirely sure what you mean by deep in a subclass, but I think the intent of the original work is quite clear. These are UCR expressions that return reporting and case sharing groups for a given user, something that is only possible for mobile workers. They abstracted out shared functionality (fetching the user, and the required methods of the UCR expression interface) into a base class for DRY reasons, but I do not believe it is in a way that obscures how things were designed to function.

Are there specific regressions you had in mind? I would be happy to try to alleviate those concerns. The only behavior change here is changing an exception that prevented insertion in the data source into returning an empty value. Even if this behavior is not a long term solution (which I do not believe is the case for the reasons mentioned above), and at some point in the future we decide to interpret these expressions in a way that can be defined for web users, I cannot see how the behavior of this PR is not strictly better than the current state.

Reply:
> I am not entirely sure what you mean by deep in a subclass

Sorry, I meant to say _base class_. Totally different meaning, sorry for the confusion.

I did some "unpacking" of the source of my concern on this PR and I think it stems from the fact that the `PillowError` records associated with this condition don't contain any tracebacks (i.e. we lack hard evidence of exactly _which_ real-world objects caused this error, and the full context that that data would provide). I generally believe (subjectively) that this fix is probably correct, but as a code reviewer, I wish there was backing data that could demonstrate this. In other words, this PR addresses a _known scenario that can break the code_, but we lack the evidence to show that _that scenario is the one that actually happened_.

In some situations it can be better to allow a bug to persist until more data can be gathered, and going by this PR alone, I can't tell if it's one of those situations or not (because I don't know what the real-world effects of this bug are, besides stacking `PillowError`s).

Reply:
> we lack hard evidence of exactly which real-world objects caused this error, and the full context that that data would provide

I am not sure I totally agree with this statement. Specifically, we have a list of exactly which cases cause this error, and have confirmed that a number of them (those I spot checked which is admittedly a small percent of the overall number of errors), raise this exception in a specific data source, `test_id`, and know exactly why (that data source passes the user that last modified the case to this expression, and for cases recently closed in a case upload that user is the web user who did the upload). We know that all cases last modified by that upload will cause the same error in the same data source. We could certainly check each case referenced to confirm they are in the same situation, but I am not sure it is fair to say we lack hard evidence, since we were able to independently reproduce this issue for each of the cases in question we tried. I guess it is possible these cases triggered the same exception elsewhere, although the remedy for that would be the same since these expressions cannot be called on a web user, but I believe we have as much evidence as any stack trace could provide, in that we have reproduced the specific exceptions for the specific cases.

Reply:
There is also evidence in Sentry: https://sentry.io/share/issue/b59cdeed4436484bace2488961267edd/ (once logged in you'll be able to see the context variables that resulted in the error).

Reply:
One last thought if you would still like to see the evidence first hand, you can do the same reproduction steps I took, trying to insert the errored documents into the relevant data source, to see this exception being raised. I am also happy to help you with that if you aren't familiar enough with this part of the code base yet.